ftw/zipextract/zipextracter.py

```diff
@@ -0,0 +1,289 @@
+from AccessControl import Unauthorized
+import os
+from plone import api
+from Products.CMFCore.interfaces import IFolderish
+from plone.i18n.normalizer.interfaces import IIDNormalizer
+from tempfile import NamedTemporaryFile
+from ftw.zipextract.interfaces import IFolderCreator
+from ftw.zipextract.interfaces import IFileCreator
+from ftw.zipextract.interfaces import IFile
+from ftw.zipextract import _
+from ftw.zipextract import logger
+import zipfile
+from ZODB.POSException import ConflictError
+from zope import component
+try:
+    from zope.app.container.interfaces import INameChooser
+except ImportError:
+    from zope.container.interfaces import INameChooser
+
+
+class _FileNode():
+
+    def __init__(self, parent_folder, filename, fileid, info):
+        self.name = _(filename)
+        self.info = info
+        self.parent_folder = parent_folder
+        self.id = fileid
+        self.is_folder = False
+
+    @property
+    def path(self):
+        return os.path.join(self.parent_folder.path, self.id)
+
+
+class _FolderNode():
+
+    def __init__(self, parent_folder, folder_name):
+        self.name = folder_name
+        self.subtree = {}
+        self.file_dict = {}
+        self.parent_folder = parent_folder
+        self.id = self.name
+        self.is_folder = True
+
+    @property
+    def path(self):
+        if self.parent_folder:
+            return os.path.join(self.parent_folder.path, self.id)
+        else:
+            return self.id
+
+    def get_files(self):
+        file_list = list(self.file_dict.values())
+        for folder_node in self.subtree:
+            file_list.extend(self.subtree[folder_node].get_files())
+        return file_list
+
+
+class ZipExtracter():
+    """
+    This should be initialized with a file object.
+    It implements an "extract" method, which creates
+    a new file object for every file in the zip, so
+    it needs to extract that file and save it to disk,
+    but also create the file object in zope
+    """
+
+    def __init__(self, context, max_size=None):
+        ifile = IFile(context)
+        if not ifile.is_zip():
+            err_msg = "{} is not a zip file, can't extract from it.".format(
+                context.absolute_url_path())
+            raise TypeError(err_msg)
+        file = ifile.get_blob()
+        self.context = context
+        self.parent_node = context.getParentNode()
+        self.zfile = zipfile.ZipFile(file.open())
+        self.file_name = os.path.basename(self.context.absolute_url_path())
+        self.file_name = os.path.splitext(self.file_name)[0]
+        self.parent_folder = os.path.dirname(self.context.absolute_url_path())
+        self.file_infos = self.extract_file_infos()
+        self._extract_file_tree()
+        self.max_size = max_size
+        self.file_creator = IFileCreator(self.context)
+
+    def _isdir(self, type):
+        if type == "Folder":
```

Comment:
`type` is a builtin function and should not be redeclared.

Reply:
Not using this class anymore, I'll delete it